book/src/proposals/booting-from-snapshot.md

```diff
@@ -0,0 +1,21 @@
+# Booting From a Snapshot
+
+## Problem
+
+A validator booting from a snapshot should know that they are caught up to the cluster before voting.
+
+## Solution
+
+A booting validator needs to have some set of trusted validators whose votes it will rely upon to ensure that it has caught up to the cluster and has a valid snapshot. With that set, the validator can
+
+1. Get a snapshot to boot from.
+
+2. Send canary transactions with its local recent blockhash to test if they get onto the ledger.
+
+3. While waiting to see one of the canary transactions, set a new root every time some threshold percent of the trusted validator stake roots a bank.
+
+4. Once a canary transaction is observed in a bank that some threshold of trusted validator stake has voted on, confirm that the `trusted validator root` > `snapshot root`
+
+5. Figure out what banks the validator is not locked out from. This is not currently done, and should be addressed separately, but is not necessary for this design.
```

Comment:
I think spec'ing out your lockouts based on your previous lock tower state after booting from a snapshot is a necessary part of the design.

I think the proper procedure here is grab the ancestry information (we store the last 10,000 parents or somethiinig) and then use that to figure out for some snapshot root `R` what the next descendant of `R` you can vote on, call this `D`. Then we wait for both

a) Catchup criteria listed above
b) See a votable bank > `D` 

before we starting voting

Reply:
If the ancestry info is coming from the validator before it stopped, that info isn't up to date. I'm not sure the 10,000 parents helps, unless it's coming from an external validator, which seems difficult to make secure.

Reply:
@TristanDebrunner I think we could make it part of the bank hash. We can use a rolling hash to keep track of the latest 10,000 parents and incorporate that in the bank hash. When we create the snapshot, we can then read out the ancestors from either blocktree/create a separate tree track the ancestors. Validators can then validate this ancestor list against the bank hash.

@sagar-solana